flow/stopwatch_dl.cc

```diff
@@ -0,0 +1,148 @@
+  auto const x = rect.x();
+  auto const y = rect.y();
+  auto const width = rect.width();
+  auto const height = rect.height();
+  auto const bottom = rect.bottom();
+
+  // Scale the graph to show time frames up to those that are 3x the frame time.
+  auto const one_frame_ms = stopwatch_.GetFrameBudget().count();
+  auto const max_interval = one_frame_ms * 3.0;
+  auto const max_unit_interval = UnitFrameInterval(max_interval);
+  auto const sample_unit_width = (1.0 / kMaxSamples);
+
+  // Provide a semi-transparent background for the graph.
+  painter.DrawRect(rect, 0x99FFFFFF);
+
+  // Prepare a path for the data; we start at the height of the last point so
+  // it looks like we wrap around.
+  {
+    for (auto i = size_t(0); i < stopwatch_.GetLapsCount(); i++) {
+      auto const sample_unit_height =
+          (1.0 - UnitHeight(stopwatch_.GetLap(i).ToMillisecondsF(),
+                            max_unit_interval));
+
+      auto const bar_width = width * sample_unit_width;
+      auto const bar_height = height * sample_unit_height;
+      auto const bar_left = x + width * sample_unit_width * i;
+
+      painter.DrawRect(SkRect::MakeLTRB(/*left=*/bar_left,
+                                        /*top=*/y + bar_height,
+                                        /*right=*/bar_left + bar_width,
+                                        /*bottom=*/bottom),
+                       0xAA0000FF);
+    }
+  }
+
+  // Draw horizontal frame markers.
+  {
+    if (max_interval > one_frame_ms) {
+      // Paint the horizontal markers.
+      auto count = static_cast<size_t>(max_interval / one_frame_ms);
+
+      // Limit the number of markers to a reasonable amount.
+      if (count > kMaxFrameMarkers) {
+        count = 1;
+      }
+
+      for (auto i = size_t(0); i < count; i++) {
+        auto const frame_height =
+            height * (1.0 - (UnitFrameInterval(i + 1) * one_frame_ms) /
+                                max_unit_interval);
+
+        // Draw a skinny rectangle (i.e. a line).
+        painter.DrawRect(SkRect::MakeLTRB(/*left=*/x,
+                                          /*top=*/y + frame_height,
+                                          /*right=*/width,
+                                          /*bottom=*/y + frame_height + 1),
+                         0xCC000000);
+      }
+    }
+  }
+
+  // Paint the vertical marker for the current frame.
+  {
+    DlColor color = DlColor::kGreen();
+    if (UnitFrameInterval(stopwatch_.LastLap().ToMillisecondsF()) > 1.0) {
+      // budget exceeded.
+      color = DlColor::kRed();
+    }
+    auto const l =
+        x + width * (static_cast<double>(stopwatch_.GetCurrentSample()) /
+                     kMaxSamples);
+    auto const t = y;
+    auto const r = l + width * sample_unit_width;
+    auto const b = rect.bottom();
+    painter.DrawRect(SkRect::MakeLTRB(l, t, r, b), color);
+  }
+
+  // Actually draw.
+  canvas->DrawVertices(painter.IntoVertices(), DlBlendMode::kSrc, paint);
```

Comment:
This is a bit odd - it looks like the colors you're using have some opacity on them. But drawing with blend mode src with blow away whatever is "under" the vertices even with opacity. I'd think you'd want to either draw 1) fully opaque and blend mode src or 2) draw partially opaque and use the default blend mode srcOver.

Reply:
It sounds like `srcOver` will be more understandable to other folks, so I'll go with that and make a note.